reckon-core/src/main/groovy/org/ajoberstar/reckon/core/InventoryService.java

```diff
@@ -0,0 +1,214 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.ajoberstar.reckon.core;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.apache.commons.lang3.builder.EqualsBuilder;
+import org.apache.commons.lang3.builder.HashCodeBuilder;
+import org.apache.commons.lang3.builder.ToStringBuilder;
+import org.apache.commons.lang3.builder.ToStringStyle;
+import org.eclipse.jgit.lib.Constants;
+import org.eclipse.jgit.lib.ObjectId;
+import org.eclipse.jgit.lib.Ref;
+import org.eclipse.jgit.lib.Repository;
+import org.eclipse.jgit.revwalk.RevCommit;
+import org.eclipse.jgit.revwalk.RevWalk;
+import org.eclipse.jgit.revwalk.RevWalkUtils;
+import org.eclipse.jgit.revwalk.filter.RevFilter;
+
+public class InventoryService {
+  private final Repository repo;
+  private final Function<Ref, Optional<ReckonVersion>> tagParser;
+
+  public InventoryService(Repository repo, Function<String, Optional<String>> tagSelector) {
+    this.repo = repo;
+    this.tagParser =
+        ref -> {
+          String tagName = Repository.shortenRefName(ref.getName());
+          return tagSelector.apply(tagName).flatMap(ReckonVersion::valueOf);
+        };
+  }
+
+  public Inventory get() {
+    try (RevWalk walk = new RevWalk(repo)) {
+      walk.setRetainBody(false);
+
+      ObjectId headObjectId = repo.getRefDatabase().getRef("HEAD").getObjectId();
+      RevCommit headCommit = walk.parseCommit(headObjectId);
+
+      Set<TaggedVersion> taggedVersions = getTaggedVersions(walk);
+
+      ReckonVersion currentVersion =
+          findCurrent(headCommit, taggedVersions.stream())
+              .map(TaggedVersion::getVersion)
+              .orElse(null);
+      TaggedVersion baseRelease =
+          findBase(walk, headCommit, taggedVersions.stream().filter(TaggedVersion::isNormal));
+      TaggedVersion baseVersion = findBase(walk, headCommit, taggedVersions.stream());
+
+      int commitsSinceBase = RevWalkUtils.count(walk, headCommit, baseVersion.getCommit());
+
+      Set<ReckonVersion> parallelVersions =
+          taggedVersions
+              .stream()
+              .map(version -> findParallel(walk, headCommit, version, taggedVersions))
+              // TODO Java 9 Optional::stream
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.ajoberstar.com/coding_rules#rule_key=squid%3AS1135)